enhancements/microshift/lvms-topolvm-downstream-replacement.md

```diff
@@ -0,0 +1,283 @@
+are still usable & upgradeable after the change to the downstream driver.
+
+As a MicroShift PM, I want to ensure that LVMS means the same feature set
+to the end user in MicroShift as it does in OpenShift.
+
+As a LVMS maintainer, I do not want to keep maintaining the bundled TopoLVM
+images only for MicroShift. I want it to switch so that maintenance and 
+support can be done solely based on the LVMS project.
+
+As a LVMS maintainer, I want to ensure that the LVMS project is used
+in as many places as possible to ensure that it is well tested and
+maintained.
+
+### Goals
+
+* Switch from using repackaged registry.redhat.io/lvms4/topolvm-rhel9 images
+  to using a minified version of LVMS in MicroShift based on registry.redhat.io/lvms4/lvms4-lvms-operator-bundle.
+* Switch from using CSI Sidecars to using the LVMS Deployment optimized for Edge.
+
+### Non-Goals
+
+* Ensuring that LVMCluster works for all use cases in MicroShift.
+  Instead, we will focus on allowing custom deviceClasses to be used 
+  outside the LVMCluster Custom Resource.
+* Integrating all usual MicroShift TopoLVM use cases into LVMS. (think of LVM RAID, etc.)
+  Instead, we will rebrand our documentation and put forward best practices 
+  for using LVMS in MicroShift based on LVMCluster with an extended documentation
+  for specialised use cases.
+* Forcing MicroShift users to migrate to LVMS CustomResources. We will allow
+  the use of the old deviceClass mechanism without a migration path and permanently
+  allow custom deviceClass management.
+
+## Proposal
+
+With 4.16 and the [controller consolitation initative in LVMS](https://issues.redhat.com/browse/OCPEDGE-689)
+we no longer use any of our own upstream images like TopoLVM or the CSI Sidecars. 
+We would like to introduce a conformant way to rebase MicroShift based on LVMS, not TopoLVM in the future.
+This will allow us to no longer ship topolvm as a component,
+but embed it into LVMS fully as well as maintain LVMS as the only
+storage driver in MicroShift and OpenShift we have to support for lvm2 support.
+
+### Workflow Description
+
+1. Edge device administrator deploys a host with MicroShift 4.15 or 4.16
+   installed, and maintains a storageClass, and deviceClass as per 
+   [the documentation reference](https://github.com/openshift/microshift/blob/main/docs/contributor/storage/configuration.md).
+2. Software runs, time passes.
+3. Edge device administrator updates the host to run MicroShift 4.17.
+4. Edge device restarts.
+5. MicroShift restarts.
+6. MicroShift removes the old TopoLVM deployment without removing the CRDs.
+   All existing storageClasses and deviceClasses are still present.
+7. MicroShift installs LVMS with a LVMCluster CustomResource.
+   * The LVMCluster is configured with `.spec.configurationPolicy=ExternalConfiguration`
+     which will trigger a deployment of the driver which can use the existing
+     deviceClasses and storageClasses.
```

Comment:
I believe this has been discussed in the last architecture call and is also (at least to the best of my knowledge) answered in https://github.com/openshift/enhancements/pull/1601/files#diff-e0dc1c4db255868fe6b65d8b8187629884953cb3900650f9b9a755fc24fcd33cR151

To your question
> /etc/microshift/lvmd.yaml is read and transposed into a LVMCluster object?

The reason that is impractical is because lvmd.yaml is the single source of truth for TopoLVM and the CSI Driver, not LVMCluster, meaning that transposing would just mean we would write them right back to lvmd.yaml on the next reconciliation. the lvmd.yaml is the user facing API of the underlying CSI Driver, LVMCluster is the user facing API of LVMS. They do overlap, but the feature set of LVMS is significantly reduced by the fact that lvcreate-options and custom volume groups are not allowed.

We (@suleymanakbas91 @brandisher @DanielFroehlich and myself) believe that doing this would mean explosion in support cases we would be needing to tackle since potentially all external volume groups and logical volumes suddenly become supported scenarios for LVMCluster.

Reply:
I guess to summarize: The only way we would be able to make this work is by

1. Creating an API Extension for LVMCluster that is covering lvcreate-options in its entirety, as well as volume group creation settings in its entirety.
2. Creating an API Extension for LVMCluster that allows for static provisioning, similar to what MicroShifts install time volume groups currently do
3. Adding Supported Scenarios / Tests for all known dynamic LVM configurations in use (LVM Software Raid being the most problematic one because it conflicts with LVM Thin Pool configurations) in both QE setup as well as E2E
4. Creating a migration tool just for MicroShift which can transition deviceClasses discovered outside of LVMCluster to be put into LVMCluster.

This was mainly rejected due to A/B, since we believe that creating a "wrapper" API around the original low-level API is not useful at best and hard to maintain at worst if we expose / transpose all settings 1:1 just into a CRD.

Reply:
> TopoLVM and the CSI Driver, not LVMCluster, meaning that transposing would just mean we would write them right back to lvmd.yaml on the next reconciliation.

Ah, that's the flaw in my logic :D Thanks for the well thought answer!